xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+// xtext.cpp
+// Lua interface to wrap a utf8 string into a XText object
+// that provides various text shaping and layout methods
+// with the help of Fribidi, Harfbuzz and libunibreak.
+
+// We do many things similarly to how they are done in crengine,
+// and took and adapted much code from it.
+// For many links and notes about the concepts and libraries used,
+// see: https://github.com/koreader/crengine/issues/307
+
+extern "C"
+{
+#include <lua.h>
+#include <lauxlib.h>
+#include <lualib.h>
+#include "xtext.h"
+}
+
+// Harfbuzz
+#include <hb.h>
+#include <hb-ft.h>
+
+// FriBiDi
+#include <fribidi/fribidi.h>
+
+// libunibreak
+#include <wordbreak.h>
+#include <linebreak.h>
+    // linebreakdef.h is not wrapped by this, unlike linebreak.h
+    // (not wrapping it results in "undefined symbol" with the
+    // original function name kinda obfuscated)
+    #ifdef __cplusplus
+    extern "C" {
+    #endif
+#include <linebreakdef.h>
+    #ifdef __cplusplus
+    }
+    #endif
+
+// luajit doesn't provide lua_set/getuservalue (unless compiled
+// with LUAJIT_ENABLE_LUA52COMPAT) but these are equivalents
+// for our purpose
+#if LUA_VERSION_NUM < 502
+#define lua_setuservalue lua_setfenv
+#define lua_getuservalue lua_getfenv
+#endif
+
+// Some names, as they should be known to Lua
+#define XTEXT_LIBNAME "xtext"
+#define XTEXT_METATABLE_NAME "luaL_XText"
+#define XTEXT_HB_FONT_DATA_METATABLE_NAME "luaL_XText_HB_Font_Data"
+#define XTEXT_LUA_HB_FONT_DATA_TABLE_KEY_NAME "_hb_font_data"
+#define XTEXT_LUA_FONT_GETFONT_CALLBACK_NAME "getFallbackFont"
+
+// Max unicode chars per shaped (visual) line
+#define MAX_LINE_CHARS 4096
+// Max returned glyphs per line (usually less glyphs than chars,
+// but allow for more just in case some fonts don't come with many
+// glyphs and combine many diacritics to form a unicode char).
+#define MAX_LINE_GLYPHS 2*MAX_LINE_CHARS
+
+// Max number of fonts (main + fallbacks)
+// (main + 15 fallback fonts should be enough)
+#define MAX_FONT_NUM 16
+
+#define NOT_MEASURED INT_MIN
```

Comment:
INT32_MIN?

Reply:
Is it different than INT_MIN ? Or it just can be, on some platforms?

Reply:
My naive assumption would be that it'll match the target's bitness (i.e., will be INT64 on 64bit), while I assume you always want an INT32.

Reply:
Okay, I get it now, see measureSegment's signature.

Reply:
Goddamnit, clicked the wrong button again. :D.

Reply:
I'm a bit confused with these `#define SOME_NUMBER`: they come with no type/length - and nothing to say in what context they should be used.
Can I assume they'll work as expected in any context?
Using `short i = INT_MIN64`, will it be cast to a short and be the min short possible?

Reply:
Unless you use a specific format identifier for literals, that'll be an `int` (which in turn will be an int32_t or an int64_t depending on arch ^^).

AFAIK, there's no identifier for "short" or "char". You can just say "unsigned" via `u` or `U`, "long" via `L` (can be specified multiple times for long long), or float via `f` (a decimal literal defaults to double).

Reply:
As for your short example, GCC will happily shout at you that it just will not do:

```
warning: overflow in conversion from ‘int’ to ‘short int’ changes value from ‘-2147483648’ to ‘0’ [-Woverflow]
  326 |   short s_min = INT_MIN;
```

Reply:
Which nicely illustrates that a literal is an `int` by default, so you get an implicit cast, which overflows ;).

Reply:
In any case, here, as it is, int makes sense, as that matches `measureSegment`'s signature.